### PR TITLE
EIP 140: adding REVERT instruction

### LINKED ISSUE #232: EIP140: REVERT
This issue keeps track of adoption of https://github.com/ethereum/EIPs/pull/206 <del>and https://github.com/ethereum/EIPs/pull/207/files</del>, which introduce the REVERT instruction.

EDIT: https://github.com/ethereum/EIPs/pull/207 is not about protocol consensus anyway, so not relevant for the Yellow Paper.

### PR DESCRIPTION
**Pull request by pirapira**
This pull-request implements https://github.com/ethereum/EIPs/pull/206 .  Merging this requires the community's consensus on the EIP.

Fixes #232

### CODE REVIEW

**Comment by pirapira** [comment]
@axic and @chriseth might be interested.

File: Paper.tex [hunk 1]
```diff
@@ -898,7 +908,7 @@ \subsubsection{Exceptional Halting}
 \end{array}
 \end{equation}
 
-This states that the execution is in an exceptional halting state if there is insufficient gas, if the instruction is invalid (and therefore its $\delta$ subscript is undefined), if there are insufficient stack items, if a {\small JUMP}/{\small JUMPI} destination is invalid or the new stack size would be larger then 1024. The astute reader will realise that this implies that no instruction can, through its execution, cause an exceptional halt.
+This states that the execution is in an exceptional halting state if there is insufficient gas, if the instruction is invalid (and therefore its $\delta$ subscript is undefined), if there are insufficient stack items, if a {\small JUMP}/{\small JUMPI} destination is invalid or the new stack size would be larger then 1024.
```
Comment (axic) [inline Paper.tex]: Unrelated to this change, but perhaps worth mentioning that:
> if the instruction is invalid (including {\small INVALID}...

File: Paper.tex [hunk 2]
```diff
@@ -767,14 +772,19 @@ \section{Message Call} \label{ch:call}
 \end{cases}
 \end{equation}
 
-The account's associated code (identified as the fragment whose Keccak hash is $\boldsymbol{\sigma}[c]_c$) is executed according to the execution model (see section \ref{ch:model}). Just as with contract creation, if the execution halts in an exceptional fashion (i.e. due to an exhausted gas supply, stack underflow, invalid jump destination or invalid instruction), then no gas is refunded to the caller and the state is reverted to the point immediately prior to balance transfer (i.e. $\boldsymbol{\sigma}$).
+The account's associated code (identified as the fragment whose Keccak hash is $\boldsymbol{\sigma}[c]_c$) is executed according to the execution model (see section \ref{ch:model}). Just as with contract creation, if the execution halts in an exceptional fashion (i.e. due to an exhausted gas supply, stack underflow, invalid jump destination or invalid instruction), then no gas is refunded to the caller and the state is reverted to the point immediately prior to balance transfer (i.e. $\boldsymbol{\sigma}$).  However, if the execution halted by a {\small REVERT} instruction, the remaining gas (if any) is paid back although the state is reverted.
```
Comment (axic) [inline Paper.tex]: Maybe we could have a proper list of "exceptional halts" somewhere.

**Comment by pirapira** [comment]
I just realized that, I still need to update the way CALL calculates its return value.  Currently it just uses the exceptional halting function `Z`.

**Comment by pirapira** [comment]
I fixed the return values of `CALL`, `CREATE` and so.

**Comment by pirapira** [comment]
@gavofyork 's suggestion was to implement this as another clause in the definition of `\Xi` (where exceptional halting is detected).  Closing for now.

**Pull request closed without merging**